src/charm.py

```diff
@@ -75,7 +75,7 @@ def _on_get_issued_certificates(self, event: ActionEvent) -> None:
         if not certificates:
             event.fail("No certificates issued yet.")
             return
-        event.set_results({"issued_certificates": certificates})
```

Comment:
What is the issue with the actions result?
If I'm not mistaken the value will be converted to string, therefore you will have Dict[str, str] from this line.

Reply:
The issue is, the automatic flattening function includes the sub-keys on the parent key. This turns the parent key into something that's not valid. This is the error that's created from the logs:

`ValueError: key ''-----BEGIN CERTIFICATE REQUEST-----\nMIIClDCCAXwCAQAwTzEeMBwGA1UEAwwVdGxzLXJlcXVpcmVyLW9wZXJhdG9yMS0w\nKwYDVQQtDCQwNzA3MTAzZS01MDIxLTQ1Y2UtOTQ5NC1jODM4OTg1NzBmMDgwggEi\nMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQC0Y/39nuAT5VD2tSVZv/YRQPlu\nA4ipB1YoBtCUgeTdNmT8d+I1T7QuwxI1cH4a+iJPkM5t9tArqg1Rz9+uHr0OltuH\nMUAg/np5uYOTj35dOmP90spvnPAAwauKwPEacJmYJo6DY9U/WN1wsd2Qg2D3nUeT\nGFqYCg25sOOVLag51LSNHgghdZxiMpdDHAOgrvqfnH0euobHNbnAJ7BUazGWbIM6\n8nMVGHP4cvzb/GD6gbs7rlDMdHBZi72riWNhGXh+mSsMxhSObHQI3SLE7NFrANWD\nVOlmQvzynIXOLZgEUlfxDmH4ZxEDohCGEqp0prKGbHeCGn+UgdHHlbnU3qIrAgMB\nAAGgADANBgkqhkiG9w0BAQsFAAOCAQEAIE7trgsb/v+GXdh3M878nB5A4kzIP4Hr\nZauPDzMXvA8KeCAS8U3dP0MDPVRahG0Xw6Ka6+xXekDuiBPfp8/iNy3tC76cslXc\nxKgp07lMi7lgLmDxpNfreSgYeHVc+hBLBCt1CcWCZGGgZrZQrE/FNuoVXjS6Sp4E\ndatnaLIPcWseeWODK1eLUMTYZu2VgzPCO6rB0xnL3wdiyVnZPaygsRzS/sp4BBgq\n/+d7gpRenryXh7Yrt5cKSlv3GvAEmJh/hDZgd+nMxJTleJZVciUEQboS17dWONEH\nl44RqOT4DPqR1JVLhdyT20Fjo5sssPSnbjpc60L/GOMtLKB/BJvtcA==\n-----END CERTIFICATE REQUEST-----'' is invalid: must be similar to 'key', 'some-key2', or 'some.key'
`

Reply:
We should probably avoid the loss of context/information caused by updating the library to flatten the value.

Instead of changing `get_issued_certificates()` to satisfy `set_results()`, we should probably format the data before supplying it to `set_results()` instead. Can we just wrap the `certificates` value in `str()`?

```python
event.set_results({"issued_certificates": str(certificates)})
```

There's probably a nicer way to do this, I haven't put much thought into it. Just trying to see if I understand the problem.

Reply:
We talked about this, and decided on doing a json dumps for each application's certificates. There was never any loss of context or information, but it's better to use a standard format like you mentioned.

Reply:
I actually agree, I'll do it that way.